### DISCUSSION TITLE
Automate user registration

### DISCUSSION THREAD

**Discussion by ForrestH** [gravitational/teleport#11814] [Q&A] [answered]
I'm trying to setup a Teleport cluster from scratch using Ansible and the only piece I cannot figure out how to automate is the creation of an admin user account. User registration requires manual intervention via the web service and my attempts to workaround this (by mimicking what the browser does using curl or inserting rows into the backend sqlite db) have not panned out so far. 

Has anyone done this before? Can we consider adding a way to automate user registration or was it omitted intentionally by design?

**Comment by webvictim** [comment]
The general idea is that users should always register themselves (because they need to configure a second factor unique to them, etc). Not permitting this to be done centrally was essentially a security decision.

With that said, it's likely possible to follow that user creation link yourself, create the user manually, do `tctl get user/<username> --with-secrets > user.yaml` to export it, then run `teleport start --bootstrap user.yaml` on another server to re-import that same user with the same password. The salt and hash are stored as part of the user record, so they can be copied between servers.

**Reply by ForrestH** [reply]
I guess I understand the reasoning to not permit this, but in my case I want to setup a shared user account where second factor authentication is tricky to share. I want it automated because we have an ansible playbook that can recreate our entire development infrastructure from scratch, so part of creating the teleport cluster would be adding an admin user account.

**Comment by ForrestH** [comment, marked as answer]
I've found a workaround that works for me:
1. Create the user with `tctl users add --roles=access,editor,auditor --logins=root admin`
2. Replace the password in the backend behind teleport's back with `sqlite3 /var/lib/teleport/backend/sqlite.db 'update kv set value="<some bcrypt hash>" where key="/web/users/admin/pwd"'`

**Reply by tuxicorn** [reply]
Where do you run the sqlite command ? I am trying to achieve the same thing but i can't replace the password

**Reply by webvictim** [reply]
On the machine where the Teleport auth server is running. This command won't work for Teleport Cloud, or a managed backend like DynamoDB.

**Reply by tuxicorn** [reply]
I can find the exact `/var/lib/teleport/backend/sqlite.db` inside the pod running the Teleport auth server but can't run any sqlite commands.

**Reply by webvictim** [reply]
Something like `echo "SELECT key FROM kv;" | sudo sqlite3 /var/lib/teleport/backend/sqlite.db` ought to work.

**Reply by rajburnwal07** [reply]
@webvictim I have tried to get into auth pod but didn't find sqlite installed. So I tried below command in order to install sqlite3 and sqlite3 command.

`kubectl --namespace teleport exec $auth_pod -- apt-get update -y && sudo apt-get install sqlite3 -y`
 
 `kubectl --namespace teleport exec $auth_pod -- sqlite3 /var/lib/teleport/backend/sqlite.db 'update kv set value="$2y$10$suzE913EfcPgwe0Mg9SALu9iSyBBiDh6KViQLak/n3npKFQBVDPsS" where key="/web/users/oklroot/pwd"'`

But its giving error "Error: unable to open database "/var/lib/teleport/backend/sqlite.db": unable to open database file"

Also when I try to exec into pod then try these command to install sqlite3 and sqlite3 update command, It works fine.

Can you help me here? what could be the issue here?

**Reply by webvictim** [reply]
@rajburnwal07 What backing storage do you use for your cluster?
